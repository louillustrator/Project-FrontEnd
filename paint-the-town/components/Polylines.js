import React from "react";
import { Platform, View } from "react-native";
import MapView from "react-native-maps";

const Polylines = ({ route }) => {
  if (Platform.OS === "android") {
    const iosRoute = getIosRoute(route);
    return (
      <View>
        {route.map((path, index) => {
        const { latLng, width, colour } = path;
        return (
          <MapView.Polyline
            key={index}
            coordinates={latLng}
            strokeWidth={width}
            strokeColor={colour}
              lineCap="round"
              linejoin="round"
            />
          );
        })}
      </View>
    );
  } else {
    const iosRoute = getIosRoute(route);
    return (
      <View>
        {iosRoute.coords.map((line, index) => {
          return (
            <MapView.Polyline
              key={index}
              coordinates={line}
              strokeWidth={iosRoute.widthMap[index]}
              strokeColors={iosRoute.colourMap[index]}
              lineCap="round"
              linejoin="round"
            />
          );
        })}
      </View>
    );
  }
};

const getIosRoute = route => {
  let coords = [],
    colourMap = [],
    widthMap = [];
  route.forEach((obj, index) => {
    if (index === 0) {
      coords.push(obj.latLng);
      let currentObjColours = [];
      for (let i = 0; i < obj.latLng.length; i++) {
        currentObjColours.push(obj.colour);
      }
      colourMap.push(currentObjColours);
      widthMap.push(obj.width);
    } else if (
      obj.latLng[0] ===
        route[index - 1].latLng[route[index - 1].latLng.length - 1] &&
      obj.width === route[index - 1].width
    ) {
      coords[coords.length - 1] = coords[coords.length - 1].concat(obj.latLng);
      let currentObjColours = [];
      for (let i = 0; i < obj.latLng.length; i++) {
        currentObjColours.push(obj.colour);
      }
      colourMap[colourMap.length - 1] = colourMap[colourMap.length - 1].concat(
        currentObjColours
      );
    } else {
      coords.push(obj.latLng);
      let currentObjColours = [];
      for (let i = 0; i < obj.latLng.length; i++) {
        currentObjColours.push(obj.colour);
      }
      colourMap.push(currentObjColours);
      widthMap.push(obj.width);
    }
  });

  return { coords, colourMap, widthMap };
};

export default Polylines;