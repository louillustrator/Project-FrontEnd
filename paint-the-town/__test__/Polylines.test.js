import React from "react";
import { Platform, View } from "react-native";
import MapView from "react-native-maps";
import { mount } from 'enzyme';
import renderer from "react-test-renderer";
import Polylines from "../components/Polylines";


describe('Polylines', () => {
    let props;
    beforeEach(() => {
        props = {
            route: []
        };
    });
    it('renders without crashing', () => {
        mount(<Polylines {...props} />);
    });
    it("renders correctly", () => {
        const tree = renderer.create(<Polylines {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("checks render for android platform", () => {
        Platform.OS = 'android';
        const wrapper = mount(<Polylines {...props} />);
        expect(wrapper.find(View)).toHaveLength(1);
    });
    it("checks render for ios platform", () => {
        Platform.OS = 'ios';
        const wrapper = mount(<Polylines {...props} />);
        expect(wrapper.find(MapView.Polyline)).toHaveLength(21);
    });
});