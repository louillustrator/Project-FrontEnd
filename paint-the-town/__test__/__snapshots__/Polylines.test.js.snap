// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polylines renders correctly 1`] = `
<View>
  <AIRMapPolyline
    index={0}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={1}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={2}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={3}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={4}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={5}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={6}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={7}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={8}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={9}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={10}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={11}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={12}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={13}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={14}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={15}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={16}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={17}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={18}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={19}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
  <AIRMapPolyline
    index={20}
    lineCap="round"
    lineJoin="round"
    strokeColor="#000"
    strokeWidth={1}
  />
</View>
`;
